### PR TITLE
Add compact items prefix in status bar

### PR DESCRIPTION
**Pull request by JasonInOttawa**
I prefer to have the progress bar next to the status bar with "Show all at once" turned on. I also like the Progress Bar to be as long as possible, to make chapter ticks easier to see.

This works mostly with the Prefix sets to Icons. This removes the icons and extra spaces around the slashes in Current Page and Pages left in book.

It makes the status bar icons that show a simple number (e.g., Pages left in chapter) ambiguous with each other.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/7127)

### CODE REVIEW

**Comment by poire-z** [comment]
Screenshot**s** (with various other settings, multi lines, etc...) please! so we can see how ugly this gets :)

**Comment by JasonInOttawa** [comment]
This is the option in Settings

Screen Shot 2021-01-09 at 3 08 36 PM

I'm running this on my laptop with the same size window in both cases below.

This is with the option turned off

Screen Shot 2021-01-09 at 3 07 59 PM

This is with the option turned on. Lots of ambiguities :)

Screen Shot 2021-01-09 at 3 08 15 PM

**Comment by poire-z** [comment]
Thanks.
Yes, not super pretty, but could make sense for people with very few items.
Mind moving the setting down in the menu , among the other more related settings, between
`Show footer separators` and `Hide empty items` - and may be rename it to `Compact items` ?

**Comment by JasonInOttawa** [comment]
Actually the wifi commit seems to have broken things :)

**Comment by JasonInOttawa** [comment]
The latest commit removes the "compact items" setting and replaces it with a prefix. 

Here is the menu:

menu

Here is what the bar looks like with "compact letters"

compact letters

Here is what is looks like with letters

letters

Here is what it looks like with icons

icons

**Comment by poire-z** [comment]
With something like this? : 
```diff
@@ -314,8 +314,9 @@ local footerTextGeneratorMap = {
     book_title = function(footer)
         local doc_info = footer.ui.document:getProps()
         if doc_info and doc_info.title then
+            local title = doc_info.title:gsub(" ", "\xC2\xA0") -- replace space with no-break-space
             local title_widget = TextWidget:new{
-                text = doc_info.title,
+                text = title,
```
and if %s strips this no break space, in your stripper, match explicite normal spaces " " instead of "%s".

**Comment by JasonInOttawa** [comment]
@poire-z Thanks very much! Changing spaces to non-break-spaces worked.

File: frontend/apps/reader/modules/readerfooter.lua
Comment (poire-z) [inline frontend/apps/reader/modules/readerfooter.lua:363]: Well, at least do it similarly as for title :/
(There, you're doing it too late, after the text has been measured and truncated - so, in case in some font a space and a no-break-space have different widths, there might be strange things.)

Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua]: @poire-z Thank you. Does this commit look better?

File: frontend/apps/reader/modules/readerfooter.lua
Comment (poire-z) [inline frontend/apps/reader/modules/readerfooter.lua:352]: Yes, better, but you can just update this one, no need for a new variable:
```lua
if chapter_title and chapter_title ~= "" then
  chapter_title = chapter_title:gsub(" ", "\xC2\xA0")
```

Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua]: True enough! I wonder if the interpreter is smart enough to optimise that out ... :)

File: frontend/apps/reader/modules/readerfooter.lua
Comment (poire-z) [inline frontend/apps/reader/modules/readerfooter.lua:1767]: No more an issue.

**Review by poire-z — APPROVED** [review]
Well, another personal tweak added to the footer - but I guess that's the footer's cross to bear :)
Anyway, quite unintruisive (with the gsub, otherwise, we'd have to tweak each of the litlle generators), so I don't mind.

File: frontend/apps/reader/modules/readerfooter.lua
Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua:86]: Just in case they want to change the letters used for Memory or Bookmarks :)

File: frontend/apps/reader/modules/readerfooter.lua
Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua]: Thanks. Took it out

File: frontend/apps/reader/modules/readerfooter.lua
Comment (poire-z) [reply frontend/apps/reader/modules/readerfooter.lua:86]: I get it - but they won't be able to :)
You need to use one of the `_(` function, like in the other stuff that you cloned:
`bookmark_count = C_("FooterCompactLetterPrefix", "B"),`

Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua]: Thanks again. I don't really know what a C_ function is :) but the new commit duplicates it from the letters. I also used the letters text verbatim (BM and M) but without the colons (they take up so much space :) )

File: frontend/apps/reader/modules/readerfooter.lua
Comment (poire-z) [inline frontend/apps/reader/modules/readerfooter.lua:86]: I think it's better to use another "context name" than `FooterLetterPrefix `, to not confuse the translators.
`FooterCompactLetterPrefix`

File: frontend/apps/reader/modules/readerfooter.lua
Comment (Frenzie) [reply frontend/apps/reader/modules/readerfooter.lua]: @poire-z Pssht. ;-P
> No more an issue.

**Comment by poire-z** [comment]
Given that in the end, you ended up using both letters and icons for prefix, you could rename the option in the menu from Compact Letters  to Compact Items, and just use for item_prefix/symbol_type : "compact" instead of "compact_letters".
We then get smaller values for the thing: letters/icons/compact.

**Comment by JasonInOttawa** [comment]
I went with item_prefix compact_items rather than compact to keep things a bit more descriptive, and perhaps to avoid name clashes later.

Just as an aside, I had to change the following in my settings.reader.lua file:

`["item_prefix"] = "compact_letters",`  

to 

`["item_prefix"] = "compact_items", 
`

or I got this exception when opening a book:

```
[!] doShowReader coroutine crashed:
frontend/apps/reader/modules/readerfooter.lua:195: attempt to index a nil value
stack traceback:
	frontend/apps/reader/modules/readerfooter.lua:1763: in function 'genFooterText'
	frontend/apps/reader/modules/readerfooter.lua:1881: in function 'updateFooterText'
	frontend/apps/reader/modules/readerfooter.lua:1861: in function 'updateFooterPage'
	frontend/apps/reader/modules/readerfooter.lua:1989: in function 'handleEvent'
	frontend/ui/widget/container/widgetcontainer.lua:88: in function 'propagateEvent'
	frontend/ui/widget/container/widgetcontainer.lua:106: in function 'handleEvent'
	frontend/ui/widget/container/widgetcontainer.lua:88: in function 'propagateEvent'
	frontend/ui/widget/container/widgetcontainer.lua:106: in function 'handleEvent'
	frontend/apps/reader/modules/readerrolling.lua:995: in function '_gotoPage'
	frontend/apps/reader/modules/readerrolling.lua:1003: in function '_gotoXPointer'
	frontend/apps/reader/modules/readerrolling.lua:183: in function 'setupXpointer'
	frontend/apps/reader/modules/readerrolling.lua:320: in function 'handleEvent'
	frontend/ui/widget/container/widgetcontainer.lua:88: in function 'propagateEvent'
	frontend/ui/widget/container/widgetcontainer.lua:106: in function 'handleEvent'
	frontend/apps/reader/readerui.lua:430: in function 'init'
	frontend/ui/widget/widget.lua:48: in function 'new'
	frontend/apps/reader/readerui.lua:578: in function 'doShowReader'
	frontend/apps/reader/readerui.lua:536: in function <frontend/apps/reader/readerui.lua:535>01/11/21-16:09:29 INFO  FM loaded plugin autosuspend at plugins/autosuspend.koplugin

```

File: frontend/apps/reader/modules/readerfooter.lua
Comment (poire-z) [inline frontend/apps/reader/modules/readerfooter.lua:86]: Ok with compact_items. So, C_("FooterCompactItemsPrefix",...) ?

**Comment by poire-z** [comment]
> I got this exception when opening a book

We might get many of this kind if we go at putting unexpected values into settings :) We don't really pay (nor need to pay) much attention to that. (but it's a bit sad when this kind of exception happens when you downgrade, and the old code can't cope with a setting that changed meaning/format/type.)
So, don't worry.

File: frontend/apps/reader/modules/readerfooter.lua
Comment (poire-z) [inline frontend/apps/reader/modules/readerfooter.lua:1770]: Put the `table.insert(info, BD.wrap(text))` outside the if/else as it's the same. And so, no `else `needed.

Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua:1770]: Should I hold off on any more commits, or keep putting them in with each comment? :)

I notice a lot of builds happening ...

Comment (Frenzie) [reply frontend/apps/reader/modules/readerfooter.lua:1770]: It's better to the environment to bundle them, but they do auto-cancel.

Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua:1770]: (I was a software designer in a prior life, but it's been 17 years. I'm rusty)

Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua:1770]: Thanks. I'll hold off committing until there are a few. Please do let me know if I'm holding anything up.

File: frontend/apps/reader/modules/readerfooter.lua
Comment (Frenzie) [inline frontend/apps/reader/modules/readerfooter.lua:83]: This probably isn't the desired character, but the one we quickly type because we don't know how to type a more suitable UTF-8 symbol.

Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua:83]: I was going for something similar to the right/left pointing arrows as in icons. I can replace them with the icon character if preferred.

File: frontend/apps/reader/modules/readerfooter.lua
Comment (poire-z) [inline frontend/apps/reader/modules/readerfooter.lua:89]: It happens you don't need `BD.mirroredUILayout() and ... or ...`, as the chars you chose are mirrorable, and so, mirrored by the FriBidi code.
You can switch the UI language to Arabic to see.
Without BD.mirroredUILayout() , you'll get > and >> mirrored just fine:
image

But there's an issue, I guess with you removing the spaces :( You don't get the expected RTL order for 1/2, which should look like 2/1:
image

Using a zero-width space instead does not seem to solve it :/
`text = text:gsub('%s', '\xE2\x80\x8B') `

File: frontend/apps/reader/modules/readerfooter.lua
Comment (Frenzie) [reply frontend/apps/reader/modules/readerfooter.lua:83]: I don't know what you're referring to precisely, but I'm talking about something like the difference between greater than and a chevron (like `⟩`).

File: frontend/apps/reader/modules/readerfooter.lua
Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua:89]: I switched to a RTL language and see this

Screen Shot 2021-01-11 at 5 04 48 PM

The spaces around the slash appear all right.

File: frontend/apps/reader/modules/readerfooter.lua
Comment (poire-z) [reply frontend/apps/reader/modules/readerfooter.lua:83]: Yes, no need for icons, there might be more accurate unicode symbols, the classic < > might not be that symetric in some fonts (or dunno, may be they are ? :)

File: frontend/apps/reader/modules/readerfooter.lua
Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua:89]: Sorry, is the issue that the page numbering should also be mirrored (showing as 683/-672 683/11)?

Comment (poire-z) [reply frontend/apps/reader/modules/readerfooter.lua:89]: The space might be, but that's not the issue:
an arabic/RTL reader expects to read page 11 of 683 as : 683/11.

Other white space that should behave:
https://www.compart.com/en/unicode/bidiclass/WS

May be use a "hair space" :) if that's not too large for you: 
`text = text:gsub("%s", "\xE2\x80\x8A")` (please use double quote " instead of single ')

image
image

Comment (poire-z) [reply frontend/apps/reader/modules/readerfooter.lua:89]: For differences in space widths: http://jkorpela.fi/chars/spaces.html

Comment (Frenzie) [reply frontend/apps/reader/modules/readerfooter.lua:89]: Wikipedia isn't horrible either. Just search for a term like "hair space" or "thin space" and you'll end up on https://en.wikipedia.org/wiki/Hair_space#Unicode

File: frontend/apps/reader/modules/readerfooter.lua
Comment (poire-z) [reply frontend/apps/reader/modules/readerfooter.lua:83]: https://shapecatcher.com/ draw > or <, and see what looks like it :) 

But I guess what you might want are the buddies of the "«" or "»" we use, but single quote.
U+203A https://www.fileformat.info/info/unicode/char/203a/index.htm

File: frontend/apps/reader/modules/readerfooter.lua
Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua:89]: I'm happy with hair spaces :)

After changing the gsub that strips whitespace only, I'm seeing this now: 683/11

Pages remaining is showing the minus sign like this: 683/671-. Should that be raised as a bug?

Comment (poire-z) [reply frontend/apps/reader/modules/readerfooter.lua:89]: No, I think `683/671-` might be just what RTL readers expect.
https://util.unicode.org/UnicodeJsps/bidi.jsp?a=-123+%2F+456&p=RTL

Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua:89]: I think so too, particularly if you consider the - to be a symbol.

**Comment by poire-z** [comment]
>  I'll hold off committing until there are a few. Please do let me know if I'm holding anything up.

That was all I guess. Go ahead.

**Comment by JasonInOttawa** [comment]
Thanks. The commit is in

File: frontend/apps/reader/modules/readerfooter.lua
Comment (poire-z) [inline frontend/apps/reader/modules/readerfooter.lua:84]: You can remove the BD.mirroredUILayout()  things left ini comments, and just add a comment after:
`pages_left = ">", -- will be auto-mirrored by BiDi`

But have you tried how https://www.fileformat.info/info/unicode/char/203a/index.htm look ?
Just copy/paste this: `›` (vs `>`)

Comment (JasonInOttawa) [reply frontend/apps/reader/modules/readerfooter.lua]: Thanks. The other character looks nice (and saves a bit of space! :) )

**Comment by JasonInOttawa** [comment]
I pushed a commit after this PR was merged and I see the changed code in koreader master. Should I revert that or something? It didn't go through review so far as I know.

I apologise for the error. I didn't realize the PR had been merged at the time of the push.

**Pull request merged**
